### PR TITLE
Migrate to C++17

### PR DESCRIPTION
**Pull request by scarf005**
#### Summary

SUMMARY: Build "Migrate to C++17"

#### Purpose of change

use new useful [C++17 features](https://github.com/AnthonyCalandra/modern-cpp-features)

#### Describe the solution

- bump ubuntu-latest to ubuntu-22.04
- bump GCC version to g++11
- bump macos version to 11
- mark C++17 warnings as warnings
- internally merged #2091 

#### Describe alternatives you've considered
merge this instead (porting to C++20 would be a better idea, tho.)
- #1624

#### Testing
compile succeeded with Makefile and Cmake.
 
#### Additional context
요약_02
500+ files compiled with 25MB+ PCH file with `-j8` or: how my SSD is wearing out faster than a speeding bullet

### CODE REVIEW

**Comment by Coolthulhu** [comment]
Clang-tidy has lots of complaints. 

```
/home/runner/work/Cataclysm-BN/Cataclysm-BN/src/achievement.cpp:381:9: error: function 'current_value' should be marked [[nodiscard]] [modernize-use-nodiscard,-warnings-as-errors]
        const cata_variant &current_value() const {
        ^
        [[nodiscard]] 
/home/runner/work/Cataclysm-BN/Cataclysm-BN/src/achievement.cpp:385:9: error: function 'requirement' should be marked [[nodiscard]] [modernize-use-nodiscard,-warnings-as-errors]
        const achievement_requirement &requirement() const {
+ '[' -f build/tools/clang-tidy-plugin/libCataAnalyzerPlugin.so ']'
        ^
+ LD_PRELOAD=build/tools/clang-tidy-plugin/libCataAnalyzerPlugin.so
+ clang-tidy -quiet tests/projectile_test.cpp
        [[nodiscard]] 
/home/runner/work/Cataclysm-BN/Cataclysm-BN/src/achievement.cpp:395:9: error: function 'ui_text' should be marked [[nodiscard]] [modernize-use-nodiscard,-warnings-as-errors]
        std::string ui_text() const {
        ^
        [[nodiscard]] 
```

```
 /home/runner/work/Cataclysm-BN/Cataclysm-BN/tools/clang-tidy-plugin/XYCheck.cpp:16:1: error: nested namespaces can be concatenated [modernize-concat-nested-namespaces,-warnings-as-errors]
namespace clang
^~~~~~~~~~~~~~~
namespace clang::tidy::cata
```

Also, mac builds are being weird:
```
Error: src/json.cpp:115:18: error: 'uncaught_exceptions' is unavailable: introduced in macOS 10.12
        && !std::uncaught_exceptions()
                 ^
```

**Comment by scarf005** [comment]
- for some reason changed `.clang-tidy` file won't be used (maybe it needs to be merged into upload)
- in `build.sh` macOS version is `10.15`, that may have to do with weirdness

I guess I'll have to manually compile clang-tidy and ensure they work.

**Comment by scarf005** [comment]
#1987 needs to be merged before this PR, since clang-tidy config uses one in `upload` branch.

File: .clang-tidy [hunk 1]
```diff
@@ -4,62 +4,68 @@
...
+  -readability-isolate-declaration,
+  -readability-magic-numbers,
+  -readability-named-parameter,
+  -readability-redundant-control-flow,
+
+  -modernize-use-nodiscard
+  -modernize-concat-nested-namespaces
+  -modernize-loop-convert
+  -modernize-unary-static-assert
+  -readability-redundant-declaration
```
Comment (olanti-p) [inline .clang-tidy:60]: The issue may be from these warnings not having commas, so tidy treats them as a single warning. Also, not sure whether YAML allows empty newlines in the middle of a multiline string.

**Review by olanti-p — COMMENTED** [review]
> (maybe it needs to be merged into upload)

Unless something's broke, GA should be running all checks on a PR using the branches' last commit as source.

File: .clang-tidy (hunk 1 above)
Comment (scarf005) [reply .clang-tidy:60]: ...oh, yeah. typos. will fix them

**Comment by scarf005** [comment]
on other thoughts, migrating #1624 for current build may be more doable.

**Comment by Coolthulhu** [comment]
Conflicts

**Comment by scarf005** [comment]
@Coolthulhu `'` to `"` error. resolved conflict now.

**Comment by olanti-p** [comment]
> C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\HostX86\x64\CL.exe /c /I..\src /I"D:\a\Cataclysm-BN\Cataclysm-BN\msvc-full-features\vcpkg_installed\x64-windows-static\x64-windows-static\include" /Zi /nologo /W1 /WX- /diagnostics:column /sdl- /MP /O2 /Oi /D _SCL_SECURE_NO_WARNINGS /D _CRT_SECURE_NO_WARNINGS /D WIN32_LEAN_AND_MEAN /D _CONSOLE /D SDL_SOUND /D TILES /D SDL_BUILDING_LIBRARY /D USE_VCPKG /D _MBCS /Gm- /EHsc /MT /GS- /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /std:c++14 /Yu"stdafx.h" /Fp"D:\a\Cataclysm-BN\Cataclysm-BN\msvc-full-features\Cataclysm-lib-vcpkg-static\Release\x64\Cataclysm-lib-vcpkg-static-Release-x64.pch" /Fo"D:\a\Cataclysm-BN\Cataclysm-BN\msvc-full-features\Cataclysm-lib-vcpkg-static\Release\x64\\" /Fd"D:\a\Cataclysm-BN\Cataclysm-BN\msvc-full-features\..\Cataclysm-lib-vcpkg-static-Release-x64.pdb" /external:W1 /Gd /TP /wd4819 /wd4146 /wd26495 /wd26444 /wd26451 /wd4068 /FIstdafx.h /FC /errorReport:queue /bigobj /utf-8 ..\src\achievement.cpp ..\src\action.cpp ..\src\active_item_cache.cpp ..\src\active_tile_data.cpp . . .

MSVC builds are still c++14

**Comment by scarf005** [comment]
that should fix it (probably)

**Comment by scarf005** [comment]
@Coolthulhu should i make vcxproj files use `crlf` in editorconfig, considering that it's auto-generated from windows?
800 line diff is purely to `crlf` -> `lf` conversion
```editorconfig
[*.vcxproj]
end_of_line = crlf
```

**Pull request merged**